src/bugpoint.rs

```diff
@@ -0,0 +1,423 @@
+//! CLI tool to reduce Cranelift IR files crashing during compilation.
+
+use crate::disasm::{PrintRelocs, PrintTraps};
+use crate::utils::{parse_sets_and_triple, read_to_string};
+use cranelift_codegen::ir::{Ebb, Function, Inst, InstBuilder, TrapCode};
+use cranelift_codegen::isa::TargetIsa;
+use cranelift_codegen::settings::FlagsOrIsa;
+use cranelift_codegen::timing;
+use cranelift_codegen::Context;
+use cranelift_reader::parse_test;
+use std::path::Path;
+use std::path::PathBuf;
+
+use indicatif::{ProgressBar, ProgressDrawTarget, ProgressStyle};
+
+pub fn run(
+    filename: &str,
+    flag_set: &[String],
+    flag_isa: &str,
+    verbose: bool,
+) -> Result<(), String> {
+    let parsed = parse_sets_and_triple(flag_set, flag_isa)?;
+    let fisa = parsed.as_fisa();
+
+    let path = Path::new(&filename).to_path_buf();
+
+    let buffer = read_to_string(&path).map_err(|e| format!("{}: {}", filename, e))?;
+    let test_file = parse_test(&buffer, None, None).map_err(|e| format!("{}: {}", filename, e))?;
+
+    // If we have an isa from the command-line, use that. Otherwise if the
+    // file contains a unique isa, use that.
+    let isa = if let Some(isa) = fisa.isa {
+        isa
+    } else if let Some(isa) = test_file.isa_spec.unique_isa() {
+        isa
+    } else {
+        return Err(String::from("compilation requires a target isa"));
+    };
+
+    std::env::set_var("RUST_BACKTRACE", "0"); // Disable backtraces to reduce verbosity
+
+    for (func, _) in test_file.functions {
+        reduce(isa, func, verbose);
+    }
+
+    Ok(())
+}
+
+enum MutationKind {
+    /// The mutation reduced the amount of instructions or ebbs.
+    Shrinked,
+    /// The mutation only changed an instruction. Performing another round of mutations may only
+    /// reduce the test case if another mutation shrank the test case.
+    Changed,
+}
+
+trait Mutator {
+    fn name(&self) -> &'static str;
+
+    fn mutation_count(&self, func: &Function) -> Option<usize>;
+
+    fn mutate(&mut self, func: Function) -> Option<(Function, String, MutationKind)>;
+
+    fn reduce(
+        &mut self,
+        isa: &TargetIsa,
+        mut func: Function,
+        progress_bar_prefix: String,
+        verbose: bool,
+        should_keep_reducing: &mut bool,
+    ) -> Function {
+        let progress =ProgressBar::with_draw_target(
+            self.mutation_count(&func).unwrap_or(0) as u64,
+            ProgressDrawTarget::stdout(),
+        );
+        progress.set_style(
+            ProgressStyle::default_bar()
+                .template("{bar:60} {prefix:40} {pos:>4}/{len:>4} {msg}"),
+        );
+
+        progress.set_prefix(&(progress_bar_prefix + &format!(" phase {}", self.name())));
+
+        for _ in 0..10000 {
+            progress.inc(1);
+
+            let (mutated_func, msg, mutation_kind) = match self.mutate(func.clone()) {
```

Comment:
Note: There is one thing which conceptually slightly troubles me a bit about this code. Here we do a `func.clone()`, but a mutator such as `RemoveInst` keep both an instruction and block index.  Technically these indexes should only be valid within a given function, and using them in a clone is abusing the entity system of Cranelift for not keeping track of its primary and secondary vectors with which they can safely be used.

While I admit that this works in practice, this could be a pain point if we were to make a more secure entity system.

Reply:
It has to clone anyway, to be able to revert mutations that are not resulting in the wanted error.